File: Tdd_CSH/Task2/LinearEquation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Task2
{
    public class LinearEquation
    {
        private List<double> coefficient = new List<double>();

        public LinearEquation(double[] coefficient)
        {
            this.coefficient = coefficient.ToList();
        }

        public LinearEquation(List<double> coefficient)
        {
            this.coefficient = coefficient.ToList();
        }
        public LinearEquation(IEnumerable<double> coefficient)
        {
            this.coefficient = coefficient.ToList();
        }

        public LinearEquation(string _coefficient)
        {
            string[] coefficient = Regex.Split(_coefficient, @"[^\d.-]");
            for (int i = 0; i < coefficient.Length; i++)
            {
                if (coefficient[i] != "")
                {
                    coefficient[i] = coefficient[i].Replace('.', ',');
                    this.coefficient.Add(double.Parse(coefficient[i]));
                }
            }
        }

        public LinearEquation(int n)
        {
            if (n > 0)
            {
                coefficient = new List<double>();
                for (int i = 0; i <= n; i++) coefficient.Add(0);
            }
            else throw new ArgumentException();
        }

        public int Size
        {
            get => coefficient.Count;
        }

        public void RandomInitialization()
        {
            Random rand = new Random();
            for (int i = 0; i < Size; i++) coefficient[i] = rand.Next(35) / 10;
        }

        public void SameInitialization(double value)
        {
            for (int i = 0; i < Size; i++) coefficient[i] = value;
        }

        public double this[int index]
        {
            get
            {
                if (index >= 0 && index < Size) return coefficient[index];
                else throw new IndexOutOfRangeException();
            }
            set
            {
                if (index >= 0 && index < Size) coefficient[index] = value;
                else throw new IndexOutOfRangeException();
            }
        }

        public override string ToString()
        {
            string result = "";
            int i;
            for (i = 0; i < Size - 2; i++)
            {
                result += (coefficient[i + 1] >= 0) ? (coefficient[i].ToString() + 'x' + (i + 1).ToString() + '+') : (coefficient[i].ToString() + 'x' + (i + 1).ToString());
            }
            result += (coefficient[i].ToString() + 'x' + (i + 1).ToString());
            result += '=' + coefficient[Size - 1].ToString();
            return result;
        }
        public static LinearEquation operator +(LinearEquation a, LinearEquation b)
        {
            var result = a.coefficient.Zip(b.coefficient, (first, second) => first + second);
            return new LinearEquation(result);
        }
        public static LinearEquation operator -(LinearEquation a, LinearEquation b)
        {
            var result = a.coefficient.Zip(b.coefficient, (first, second) => first - second);
            return new LinearEquation(result);
        }

        public static LinearEquation operator *(LinearEquation a, double r)
        {
            List<double> result = new List<double>();
            for (int i = 0; i < a.Size; i++)
                result.Add(a[i] * r);
            return new LinearEquation(result);
        }

        public static LinearEquation operator -(LinearEquation a)
        {
            List<double> result = new List<double>();
            for (int i = 0; i < a.Size; i++) result.Add(-a[i]);
            return new LinearEquation(result);
        }

        public static LinearEquation operator *(double r, LinearEquation a)
        {
            return a * r;
        }

        public static bool operator ==(LinearEquation a, LinearEquation b)
        {
            return a.Equals(b);
        }

        public static bool operator !=(LinearEquation a, LinearEquation b)
        {
            return !(a == b);
        }

        public static implicit operator double[](LinearEquation a)
        {
            return a.coefficient.ToArray();
        }

        public static bool operator false(LinearEquation a)
        {
            if (a) return false;
            else return true;
        }

        public static bool operator true(LinearEquation a)
        {
            for (int i = 0; i < a.Size - 1; i++)
                if (a[i] != 0) return true; 
            return (a[a.Size - 1] != 0) ? false : true;
        }

        public void Copy(LinearEquation b)
        {
            b.coefficient = coefficient.ToList();
        }

        public bool IsNull()
        {
            for (int i = 0; i < Size; i++)
                if (this[i] != 0) return false;
            return true;
        }

        public override bool Equals(object obj)
        {
            LinearEquation b = (LinearEquation)obj;
            for (int i = 0; i < Size; i++)
            {
                if (Math.Abs(this[i] - b[i]) > 1e-9) return false;
            }
            return true;
        }
    }
}

File: Tdd_C++/Task2/LinearEquation.cpp
#include "LinearEquation.h"
#include<ctime>
#include<iostream>
#include<stdexcept>
#include<regex>
#include<list>
using namespace std;

vector<string> LinearEquation::resplit(const string& s, string rgx_str) {
	vector<string> element;
	regex Regex(rgx_str);
	sregex_token_iterator iterator(s.begin(), s.end(), Regex, -1);
	std::sregex_token_iterator end;
	while (iterator != end) {
		element.push_back(*iterator);
		++iterator;
	}
	return element;
}
LinearEquation::LinearEquation(string _s)
{
	vector<string> s = resplit(_s, "[^\\d-.]");
	for (int i = 0; i < s.size(); i++)
		if (s[i] != "") coefficients.push_back(stod(s[i].c_str()));
}
LinearEquation::LinearEquation(list<double>a)
{
	copy(a.begin(), a.end(), back_inserter(coefficients));
}
LinearEquation::LinearEquation(vector<double>a)
{
	coefficients = a;
}
LinearEquation::LinearEquation(int n)
{
	coefficients.resize(n + 1);
}

void LinearEquation::random_initialization()
{

	for_each(coefficients.begin(), coefficients.end(), [](double& t) {t = (rand() % 100) / 10.0; });
}

void LinearEquation::same_initialization(double value)
{
	for_each(coefficients.begin(), coefficients.end(), [value](double& t) {t = value; });
}
double& LinearEquation::operator[](int index)
{
	if (index >= 0 && index < coefficients.size())
		return coefficients.at(index);
	else
		throw std::out_of_range("Выход за границы.");
}
LinearEquation LinearEquation::operator*(const double& k)
{
	vector<double> result = coefficients;
	for_each(result.begin(), result.end(), [k](double& t) {t *= k; });
	return LinearEquation(result);
}

LinearEquation operator*(double k, LinearEquation& a)
{
	return a * k;
}

LinearEquation LinearEquation::operator+(LinearEquation& b)
{
	vector<double> result = coefficients;
	for (int i = 0; i < size(); i++)
		result[i] += b[i];
	return LinearEquation(result);
}

LinearEquation LinearEquation::operator-(LinearEquation& b)
{
	vector<double> result = coefficients;
	for (int i = 0; i < size(); i++)
		result[i] -= b[i];
	return LinearEquation(result);
}
LinearEquation LinearEquation::operator-()
{
	vector<double> result = coefficients;
	for_each(result.begin(), result.end(), [](double& t) {t = -t; });
	return LinearEquation(result);
}
LinearEquation::operator string()
{
	string result = "";
	int i;
	for (i = 0; i < size() - 2; i++)
	{
		result += (coefficients[i + 1] >= 0) ? (to_string(coefficients[i]) + 'x' + to_string((i + 1)) + '+') : (to_string(coefficients[i]) + 'x' + to_string((i + 1)));
	}
	result += (to_string(coefficients[i]) + 'x' + to_string((i + 1)));
	result += '=' + to_string(coefficients[size() - 1]);
	return result;
}
LinearEquation::operator bool()
{
	for (int i = 0; i < size() - 1; i++)
		if (coefficients[i] != 0) return true;
	return (coefficients[size() - 1] != 0) ? false : true;
}

LinearEquation::operator list<double>()
{
	list<double> coefficient;
	copy(coefficients.begin(), coefficients.end(), back_inserter(coefficient));
	return coefficient;
}
bool LinearEquation::isNull()
{
	for (int i = 0; i < size(); i++)
		if (coefficients[i] != 0)
			return false;
	return true;
}
bool operator==(LinearEquation& a, LinearEquation& b)
{
	for (int i = 0; i < a.size(); i++) {
		if (abs(a[i] - b[i]) > 1e-9)
			return false;
	}
	return true;
}

bool operator!=(LinearEquation& a, LinearEquation& b)
{
	return !(a == b);
}